File: Linked_Lists/go/loop_detection.go
package main

import (
  "types"
  "fmt"
)


func detect(node1 *types.Node) *types.Node {
  node2 := node1.Next

  for node1 != nil {
    if node1 == node2 {
      return node1
    }

    node1 = node1.Next
    node2 = node2.Next.Next
  }

  return &types.Node{}
}


func main() {
  var list types.LinkedList
  var loopBegin *types.Node

  list.Append(1)
  list.Append(2)
  list.Append(3)
  list.Append(4)
  node := list.Head


  for node.Next != nil {
    node = node.Next
    loopBegin = node
  }

  list.Append(5)
  list.Append(6)
  list.Append(7)

  node = list.Head

  for node.Next != nil {
    node = node.Next
  }

  node.Next = loopBegin

  fmt.Println(detect(list.Head))
}

File: Linked_Lists/python/loop_detection.py
from LinkedList import LinkedList


def detect(node1):
    node2 = node1.next

    while node1 is not None:
        if node1 == node2:
            return node1

        node1 = node1.next
        node2 = node2.next.next

    return None


def main():
    linked_list = LinkedList()
    linked_list.append(1)
    linked_list.append(2)
    linked_list.append(3)
    linked_list.append(4)
    node = linked_list.head

    while node.next is not None:
        node = node.next
        loop_begin = node

    linked_list.append(5)
    linked_list.append(6)
    linked_list.append(7)

    node = linked_list.head

    while node.next is not None:
        node = node.next

    node.next = loop_begin

    print(detect(linked_list.head))


if __name__ == '__main__':
    main()
